atc/syslog/drainer.go

```diff
@@ -39,61 +40,81 @@ func (d *drainer) Run(ctx context.Context) error {
 
 	builds, err := d.buildFactory.GetDrainableBuilds()
 	if err != nil {
-		logger.Error("Syslog drainer getting drainable builds error.", err)
+		logger.Error("failed-to-get-drainable-builds", err)
 		return err
 	}
 
 	if len(builds) > 0 {
 		syslog, err := Dial(d.transport, d.address, d.caCerts)
 		if err != nil {
-			logger.Error("Syslog drainer connecting to server error.", err)
+			logger.Error("failed-to-connect", err)
 			return err
 		}
-		defer syslog.Close()
+
+		// ignore any errors coming from syslog.Close()
+		defer db.Close(syslog)
 
 		for _, build := range builds {
-			events, err := build.Events(0)
+			err := d.drainBuild(logger, build, syslog)
 			if err != nil {
-				logger.Error("Syslog drainer getting build events error.", err)
 				return err
 			}
+		}
+	}
+	return nil
+}
+
+func (d *drainer) drainBuild(logger lager.Logger, build db.Build, syslog *Syslog) error {
+	logger = logger.Session("drain-build", lager.Data{
+		"team":     build.TeamName(),
+		"pipeline": build.PipelineName(),
+		"job":      build.JobName(),
+		"build":    build.Name(),
+	})
 
-			for {
-				ev, err := events.Next()
-				if err != nil {
-					if err == db.ErrEndOfBuildEventStream {
-						break
-					}
-					logger.Error("Syslog drainer getting next event error.", err)
-					return err
-				}
-
-				if ev.Event == "log" {
-					var log event.Log
-
-					err := json.Unmarshal(*ev.Data, &log)
-					if err != nil {
-						logger.Error("Syslog drainer unmarshalling log error.", err)
-						return err
-					}
-
-					payload := log.Payload
-					tag := build.TeamName() + "/" + build.PipelineName() + "/" + build.JobName() + "/" + build.Name() + "/" + string(log.Origin.ID)
-
-					err = syslog.Write(d.hostname, tag, time.Unix(log.Time, 0), payload)
-					if err != nil {
-						logger.Error("Syslog drainer sending to server error.", err)
-						return err
-					}
-				}
+	events, err := build.Events(0)
+	if err != nil {
+		return err
+	}
+
+	// ignore any errors coming from events.Close()
+	defer db.Close(events)
+
+	for {
+		ev, err := events.Next()
+		if err != nil {
+			if err == db.ErrEndOfBuildEventStream {
+				break
+			}
+			logger.Error("failed-to-get-next-event", err)
+			return err
+		}
+
+		if ev.Event == event.EventTypeLog {
+			var log event.Log
+
+			err := json.Unmarshal(*ev.Data, &log)
+			if err != nil {
+				logger.Error("failed-to-unmarshal", err)
+				return err
 			}
 
-			err = build.SetDrained(true)
+			payload := log.Payload
+			tag := build.TeamName() + "/" + build.PipelineName() + "/" + build.JobName() + "/" + build.Name() + "/" + string(log.Origin.ID)
+
+			err = syslog.Write(d.hostname, tag, time.Unix(log.Time, 0), payload)
 			if err != nil {
-				logger.Error("Syslog drainer setting drained on build error.", err)
+				logger.Error("failed-to-write-to-server", err, lager.Data{"tag": tag})
```

Comment:
Hey,

I have the impression that passing the `tag` there in the context of the log might end up being a duplication of what's already in the logging session (given that the session already contains pretty much all of the fields used to form the tag).

Wdyt of getting rid of that?

Reply:
Whoops, must've missed it!